_posts/2022-06-08-packaging-functions-with-helm.md

```diff
@@ -0,0 +1,221 @@
+As far as we know, the most common way to deploy functions right now is by using the `faas-cli` commands. The CLI uses the OpenFaaS REST API to deploy functions. In modern production environments you might want to use a GitOps approach to deploy build artefacts from your Continuous Integration pipeline to your cluster. At this time, two of the most popular open-source projects for GitOps are [ArgoCD](https://argo-cd.readthedocs.io/en/stable/) and [Flux](https://fluxcd.io/). Both of these projects are designed to apply Kubernetes YAML files and can not directly interact with the OpenFaaS REST API. As an alternative to deploying functions through the REST API, OpenFaaS supports a second op mode using a Kubernetes [Custom Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+![Using the CRD to deploy functions](/images/2021-06-kubectl-functions/operator-crd.png)
+> Using the CRD to deploy functions
+
+## Deploying functions using the Function CRD
+We will look into two different approaches to deploy functions when using OpenFaaS with the Function Custom Resource Definition and Operator mode. The first method will use the `faas-cli` to convert the OpenFaaS stack.yaml file into a Kubernetes CRD YAML file. The resulting files can be deployed using `kubectl` or committed to a repo where GitOps tools like [ArgoCD](https://argo-cd.readthedocs.io/en/stable/) or [Flux](https://fluxcd.io/) could use them to deploy the functions. In the second approach the functions are packaged as a Helm Chart. Helm can then be used to deploy the functions and provide configuration values to the deployment. Both are valid options and they come with their own benefits and drawbacks.
+
+CRD files:
+- CRDs can be generated from a stack.yaml file using `faas-cli generate`
+- Basic templating support using [YAML environment variable substitution](https://docs.openfaas.com/reference/yaml/#yaml-environment-variable-substitution)
+
+Helm Chart:
+- Once templates are generated you can not just run `faas-cli generate` again to reflect changes in your stack.yaml file.
+- Because Helm uses the [Go template language](https://pkg.go.dev/text/template) it supports advanced templating (logic, functions).
+- Values are centralised in the values.yaml file.
+- Charts can be versioned and Helm provides mechanisms to easily upgrade, rollback and uninstall said Charts.
+- Helm provides you the option to group various Kubernetes components together with your functions into a single deployable unit.
+
+Using Helm to package functions does not mean you wont need the YAML environment variable substitution anymore. While the Helm templates contain the deployment information, the stack.yaml file still contains the build configuration for functions. Environment variable substitution can be used to provide different configurations for building functions locally vs building and pushing functions in a CI environment.
+
+## Generating CRD files
```

Comment:
Explain that the user can also get any function deployed via faas-cli or kubectl again with:

```
kubectl get -n openfaas-fn function/name -o yaml
```

Reply:
I'm not sure I understand what you want here.

Reply:
Run the command in your cluster and see what it does.